### PR TITLE
Add integration tests for slurmctld operator

### PR DESCRIPTION
**Pull request by NucciTheBoss**
## Description

This pull request adds integration tests for the slurmctld operator. The tests first build and deploy slurmctld against the other slurm charms from latest/edge, checks that the slurmctld, munge, and etcd daemons are up, and verifies that slurmctld is listening on port 6817. 

## How was the code tested?

I ran the integration tests locally on my workstation.

## Related issues and/or tasks

This pull request is a part of the Charmed SLURM project.

## Checklist

- [x] I am the author of these changes, or I have the rights to submit them.
- [x] I have added the relevant changes to the README and/or documentation.
- [x] I have self reviewed my own code.
- [ ] All requested changes and/or review comments have been resolved.

### CODE REVIEW

**Comment by NucciTheBoss** [comment]
You will need to test my changes locally since the slurmctld charm is still failing the lint tests. If they fail both unit and lint tests, I do not want to waste time running integration tests through GitHub Actions. Next pull request will be focusing on getting the code to pass lint rules (need to agree on a standard with Omnivector.)

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -1,4 +1,7 @@
-name: slurmdbd charm tests
+# Copyright 2023 Canonical Ltd.
+# Apache Software License, version 2.0
```
Comment (dvdgomez) [inline .github/workflows/ci.yaml]: Is this change necessary? I'd like to stay as similar to existing charms as possible ie `# See LICENSE file for licensing details.`

File: tests/integration/conftest.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# Apache Software License, version 2.0
```
Comment (dvdgomez) [inline tests/integration/conftest.py]: See previous comment on License

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        logger.info(f"Setting resource {VERSION} to value {VERSION_NUM}...")
+        version.write_text(VERSION_NUM)
+    if not (etcd := pathlib.Path(ETCD)).exists():
+        logger.info(f"Getting resource {ETCD} from {ETCD_URL}...")
+        request.urlretrieve(ETCD_URL, etcd)
+
+    return {"etcd": etcd}
+
+
+def get_slurmd_res() -> Dict[str, pathlib.Path]:
```
Comment (dvdgomez) [inline tests/integration/helpers.py:50]: In order to deploy slurmd, successfully reach active status, only nhc is needed as a resource. Unless your tests specifically need singularity I think you can skip it.

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        ),
+        ops_test.model.deploy(
+            SLURMD,
+            application_name=SLURMD,
+            channel="edge",
+            num_units=1,
+            resources=slurmd_res,
+            series=series,
+        ),
+        ops_test.model.deploy(
```
Comment (dvdgomez) [inline tests/integration/test_charm.py:64]: Adjust formatting here

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+            SLURMDBD, application_name=SLURMDBD, channel="edge", num_units=1, series=series
+        ),
+        ops_test.model.deploy(
```
Comment (dvdgomez) [inline tests/integration/test_charm.py:71]: Adjust formatting here

File: tests/integration/helpers.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        logger.info(f"Getting resource {SINGULARITY_DEB} from {SINGULARITY_DEB_URL}...")
+        request.urlretrieve(SINGULARITY_DEB_URL, singularity_deb)
+    if not (singularity_rpm := pathlib.Path(SINGULARITY_RPM)).exists():
+        logger.info(f"Getting resource {SINGULARITY_RPM} from {SINGULARITY_RPM_URL}...")
+        request.urlretrieve(SINGULARITY_RPM_URL, singularity_rpm)
+
+    return {"nhc": nhc, "singularity-deb": singularity_deb, "singularity-rpm": singularity_rpm}
+
+
+@contextlib.asynccontextmanager
```
Comment (dvdgomez) [inline tests/integration/helpers.py]: This seems like a lot to test one command inside the unit. I think you could save yourself some trouble by using subprocess as seen here: https://github.com/canonical/kafka-operator/blob/84762f1d5af5b32d75a2dcb61e6312976bc42b06/tests/integration/helpers.py#L28

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+@tenacity.retry(
+    wait=tenacity.wait.wait_exponential(multiplier=2, min=1, max=30),
+    stop=tenacity.stop_after_attempt(10),
+    reraise=True,
+)
+async def test_slurmctld_port_listen(ops_test: OpsTest) -> None:
+    """Test that slurmctld is listening on port 6817."""
+    logger.info("Checking that slurmctld is listening on port 6817...")
+    async with unit_connection(ops_test, SLURMCTLD, UNIT) as conn:
+        conn.exec_command("sudo apt-get -y install net-tools")
```
Comment (dvdgomez) [inline tests/integration/test_charm.py]: Hmm, can't you just use requests as seen here?: https://github.com/jnsgruk/parca-operator/blob/2b8a98a89101f67a7222d26ed72679cb7351c097/tests/integration/test_charm.py#L34

**Review by dvdgomez — CHANGES_REQUESTED** [review]
Great start to the integration tests! I think this is definitely on the right track

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (NucciTheBoss) [reply .github/workflows/ci.yaml]: I have it as `Apache Software License, version 2.0` because we cannot guarantee that the license of the Omnivector charms will stay the same in the future. Let's say that a few years down the line, whoever owns the slurm charms decides to change the license from Apache-2.0 to AGPLv3. If we have it as `See LICENSE file for licensing details`, that will effectively convince other devs that the file is licensed under AGPLv3 when it is actually licensed under Apache-2.0 by Canonical.

Perhaps @wolsen knows what format we should use for the license header in files contributed by Canonical to an external project?

File: tests/integration/conftest.py (hunk 2 above)
Comment (NucciTheBoss) [reply tests/integration/conftest.py]: See response in https://github.com/canonical/slurmctld-operator/pull/1#discussion_r1084252362

File: tests/integration/helpers.py (hunk 3 above)
Comment (NucciTheBoss) [reply tests/integration/helpers.py:50]: Let me test that in a new commit.

File: tests/integration/test_charm.py (hunk 4 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py:64]: This formatting was generated by black. If I fix the formatting manually, black might revert it back. Let me see what happens when I play with the formatting a bit.

File: tests/integration/test_charm.py (hunk 5 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py:71]: See response here https://github.com/canonical/slurmctld-operator/pull/1#discussion_r1084256422

File: tests/integration/test_charm.py (hunk 7 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py]: Let me try that.

File: tests/integration/helpers.py (hunk 6 above)
Comment (NucciTheBoss) [reply tests/integration/helpers.py]: I assume that you are talking about the asynchronous context manager? If so, I find that it's just another way to peel the same potato. My other reasoning for using this context manager is that if I am testing locally (my workstation rather than GitHub Actions), I am using temporary ssh keys rather than my main ssh key on my system. Also, I just like it because you can control the implementation + you can chain/pipe commands and use wildcards inside the unit rather on the host system.

I think the asynchronous context manager will be useful to have in the future (especially for complicated tests), but I would like to hear your thoughts further.

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (NucciTheBoss) [reply .github/workflows/ci.yaml]: Should use the full-header here mentioned on the Apache website: https://www.apache.org/licenses/LICENSE-2.0#apply

File: tests/integration/test_charm.py (hunk 7 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py]: Hmm... After some further investigation, I do not believe we can just use the _requests_ library instead of what I have here. Reason saying is that slurmctld does not use HTTP to communicate with the other SLURM services. `requests.get()` is a wrapper for issuing a GET command to an HTTP service. It wouldn't work for checking if the slurmctld port is active since SLURM just uses TCP.

Originally I did try to use `lsof` instead of `netstat`, but I could not get it to work the way I wanted to. `lsof` comes by default inside the Juju units, so let me see if I can get the command to work again. If so, I should be able to remove the install command for netstat

File: tests/integration/helpers.py (hunk 6 above)
Comment (NucciTheBoss) [reply tests/integration/helpers.py]: Another thing. When testing locally, my ssh key set up with Juju requires a password. This will bork the automatic flow of the tests.

File: tests/integration/conftest.py (hunk 2 above)
Comment (NucciTheBoss) [reply tests/integration/conftest.py]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/6fa8296081749b73897e3c2c72bbff5737f99322

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (NucciTheBoss) [reply .github/workflows/ci.yaml]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/6fa8296081749b73897e3c2c72bbff5737f99322

File: tests/integration/test_charm.py (hunk 4 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py:64]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/8c5b6139462e9db40ec1096f09d8c0f8d7d02b94

File: tests/integration/test_charm.py (hunk 5 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py:71]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/8c5b6139462e9db40ec1096f09d8c0f8d7d02b94

File: tests/integration/test_charm.py (hunk 7 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/8c5b6139462e9db40ec1096f09d8c0f8d7d02b94. Got the test down to four lines!

File: tests/integration/helpers.py (hunk 3 above)
Comment (NucciTheBoss) [reply tests/integration/helpers.py:50]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/53dcb1cc224ebcaf1e8a0b4dd66a38988ae2c66e

File: tests/integration/test_charm.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+
+@pytest.mark.abort_on_fail
+@tenacity.retry(
+    wait=tenacity.wait.wait_exponential(multiplier=2, min=1, max=30),
+    stop=tenacity.stop_after_attempt(3),
+    reraise=True,
+)
+async def test_munge_is_active(ops_test: OpsTest) -> None:
+    """Test that slurmctld is active inside Juju unit."""
```
Comment (dvdgomez) [inline tests/integration/test_charm.py]: Looks like this needs to be updated for munge

Comment (NucciTheBoss) [reply tests/integration/test_charm.py]: Rookie mistake :clown_face:

File: tests/integration/helpers.py (hunk 6 above)
Comment (NucciTheBoss) [reply tests/integration/helpers.py]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/975e4a3d1548d6df577d08572406eb25c07639d9. See comment here for link to the public asynchronous context manager gist: https://github.com/canonical/slurmdbd-operator/pull/4#discussion_r1085617719.

File: tests/integration/test_charm.py (hunk 8 above)
Comment (NucciTheBoss) [reply tests/integration/test_charm.py]: Resolved in https://github.com/canonical/slurmctld-operator/pull/1/commits/975e4a3d1548d6df577d08572406eb25c07639d9

**Review by dvdgomez — APPROVED** [review]
Awesome, great work LGTM!

**Comment by NucciTheBoss** [comment]
> Awesome, great work LGTM!

Nice, @wolsen do you want to review before David merges?
